ENH: `signal.{envelope,resample,resample_poly}`: array API standard support

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

towards https://github.com/scipy/scipy/issues/20678

#### What does this implement/fix?
<!--Please explain your changes.-->

Convert `envelope`, `resample` and `resample_poly` to be array API compatible.

- [x] envelope
- [x] resample
- [x] resample_poly

This completes the conversion of the 'filtering' rubic of scipy.signal (https://docs.scipy.org/doc/scipy/reference/signal.html#filtering)

While at it, convert

- [x] remez
- [x] firwin
- [x] firwin2
- [x] firls
- [x] minimum_phase

This completes `tests/test_fir_filter_design.py`

ev-br commented:
Have to add quite a few skips for jax. The operations not covered by the (already horrid) `xpx.at(...).set(...)` pattern include in-place multiplies and additions: `Y[tuple(sl)] *= 0.5`,  `Y[tuple(sl)] += X[tuple(sl)]`, `X *= xp.reshape(W, newshape_W)` etc etc. I suppose it's best we add skips now and improve at leisure when jax support matures and there's dedicate bandwidth for tackling it.

lucascolley commented:
> in-place multiplies and additions

These are covered! https://data-apis.org/array-api-extra/generated/array_api_extra.at.html

ev-br commented:
> These are covered! https://data-apis.org/array-api-extra/generated/array_api_extra.at.html

Good to know. Let's keep these to a separate PR though, this one is ugly enough already:
Since `resample` uses `firwin`, convert `firwin` and its FIR filter design ilk in this PR (updated).

ev-br commented:
This one is ready from my side.

lucascolley commented on scipy/signal/_delegators.py:
```diff
@@ -334,7 +338,11 @@ def firls_signature(numtaps, bands, desired, *, weight=None, fs=None):
 def firwin_signature(numtaps, cutoff, *args, **kwds):
-    return array_namespace(cutoff)
+    if isinstance(cutoff, (int, float)):
+        xp = array_namespace(np.empty(1))   # np_compat
```
optional: could just import `np_compat` from `_lib._array_api`

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -3618,24 +3629,28 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
         if callable(window):
-            W = window(sp_fft.fftfreq(Nx))
-        elif isinstance(window, np.ndarray):
+            # older numpies do not have the .device attribute
+            device_arg = {'device': x.device if hasattr(x, 'device') else None}
+            W = window(sp_fft.fftfreq(Nx, xp=xp, **device_arg))
```
please use `xp_device` from `_lib._array_api` instead of these extra lines. The same pattern is repeated below also.

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -3618,24 +3629,28 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
+            W = window(sp_fft.fftfreq(Nx, xp=xp, **device_arg))
+        elif hasattr(window, 'shape'):   # must be an array object
             if window.shape != (Nx,):
                 raise ValueError('window must have the same length as data')
             W = window
         else:
-            W = sp_fft.ifftshift(get_window(window, Nx))
+            # older numpies do not have the .device attribute
+            device_arg = {'device': x.device if hasattr(x, 'device') else None}
+            W = sp_fft.ifftshift(get_window(window, Nx, xp=xp, **device_arg))
 
         newshape_W = [1] * x.ndim
         newshape_W[axis] = X.shape[axis]
         if real_input:
             # Fold the window back on itself to mimic complex behavior
-            W_real = W.copy()
-            W_real[1:] += W_real[-1:0:-1]
+            W_real = xp.asarray(W, copy=True)
+            W_real[1:] += xp.flip(W_real[1:])
```
I should come back to this to convince myself that these are equivalent

DietBru replied:
It is :smiley::
```python
>>> x = np.arange(5) + 10
>>> y0, y1, y2 = (np.flip(x[1:]), x[-1:0:-1], x[:0:-1])
>>> all(y1 == y0), all(y2 == y0)      
(True, True)
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -3841,45 +3858,50 @@ def resample_poly(x, up, down, axis=0, window=('kaiser', 5.0),
+                   xp.zeros(n_post_pad, dtype=h.dtype)))
     n_pre_remove_end = n_pre_remove + n_out
 
+    def _median(x, *args, **kwds):
+        return xp.asarray(np.median(np.asarray(x), *args, **kwds))
```
discussion of this from Athan at https://github.com/data-apis/array-api/issues/795#issuecomment-2090852426

ev-br replied:
Yeah, we'll be able to update or remove the shim when the upstream discussion concludes, I think.

lucascolley replied:
https://github.com/data-apis/array-api/issues/795#issuecomment-2695726917

can we use `stats.quantile`?

ev-br replied:
Let's mark it for a potential follow-up. It'll need making sure that there is no perf hit on regular numpy usage.

lucascolley commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -372,25 +374,31 @@ def firwin(numtaps, cutoff, *, width=None, window='hamming', pass_zero=True,
 
     """
+    if isinstance(cutoff, (int, float)):
+        xp = array_namespace(np.empty(1))   # np_compat
```
see other comment

lucascolley commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -836,8 +858,9 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
     bands = np.asarray(bands).copy()
-    return _sigtools._remez(numtaps, bands, desired, weight, tnum, fs,
-                            maxiter, grid_density)
+    result =  _sigtools._remez(numtaps, bands, desired, weight, tnum, fs,
+                               maxiter, grid_density)
```
```suggestion
    result = _sigtools._remez(numtaps, bands, desired, weight, tnum, fs,
                              maxiter, grid_density)
```

lucascolley commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -1231,13 +1258,17 @@ def minimum_phase(h: np.ndarray,
-    if not np.allclose(h[-n_half:][::-1], h[:n_half]):
+    n_half = h.shape[0] // 2
+
+    if not xp.any(xp.flip(h[-n_half:]) - h[:n_half] <= 1e-8 + 1e-6*abs(h[:n_half])):
+   # if not np.allclose(h[-n_half:][::-1], h[:n_half]):
```
I should also come back and convince myself that these are equivalent

lucascolley reviewed (commented):
I haven't looked at the test changes yet, but looks good so far, thanks!

j-bowhay commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -426,26 +434,27 @@ def firwin(numtaps, cutoff, *, width=None, window='hamming', pass_zero=True,
     for left, right in bands:
-        h += right * sinc(right * m)
-        h -= left * sinc(left * m)
+        h += right * xp.sinc(right * m)
```
```suggestion
        h += right * xpx.sinc(right * m, xp=xp)
```
Do we need to use xpx here?

j-bowhay commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -622,30 +634,34 @@ def firwin2(numtaps, freq, gain, *, nfreqs=None, window='hamming',
-        eps = np.finfo(float).eps * nyq
-        for k in range(len(freq) - 1):
+        freq = xp.asarray(freq, copy=True)
+        eps = xp.finfo(xp_default_dtype(xp)).eps * nyq
```
Seem odd that this isn't using `freq.dtype` but it matches the original so :shrug:

ev-br replied:
Agreed on both counts :-). And indeed, matching the original is what is meant to be in scope here.

j-bowhay commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -1231,13 +1258,17 @@ def minimum_phase(h: np.ndarray,
+    if xp.isdtype(h.dtype, "complex floating"):
         raise ValueError('Complex filters not supported')
-    if h.ndim != 1 or h.size <= 2:
+    if h.ndim != 1 or xp_size(h) <= 2:
```
Nit `h` is 1d
```suggestion
    if h.ndim != 1 or h.shape[0] <= 2:
```

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -4676,7 +4702,7 @@ def filtfilt(b, a, x, axis=-1, padtype='odd', padlen=None, method='pad',
 
     """
-    xp = array_namespace(b, a, x)
+    xp = array_namespace(_skip_if_float(b), _skip_if_float(a), x)
```
Again possibly no longer needed #22422

ev-br commented on scipy/signal/_delegators.py:
```diff
@@ -305,13 +305,17 @@ def detrend_signature(data, axis=1, type='linear', bp=0, *args, **kwds):
 
 def filtfilt_signature(b, a, x, *args, **kwds):
-    return array_namespace(b, a, x)
+    return array_namespace(_skip_if_float(b), _skip_if_float(a), x)
```
Indeed, done. Here and elsewhere.

ev-br commented:
Review comments addressed.

j-bowhay commented on scipy/signal/tests/test_fir_filter_design.py:
```diff
@@ -162,63 +182,83 @@ def test_highpass(self):
         kwargs = dict(cutoff=[0.3, 0.7], window=('kaiser', beta), scale=False)
         taps = firwin(ntaps, pass_zero=False, **kwargs)
 
+        ns = array_namespace(np.empty(0))
```
Is there a reason we can't pass `np_compat` directly?

ev-br commented:
Something bad goes on in dask, where  `37 failed, 76234 passed `.

lucascolley reviewed (commented):
```
scipy/signal/_signaltools.py:4232: in lfilter_zi
    xp = array_namespace(b, a)
        a          = array([1.])
        b          = dask.array<array, shape=(31,), dtype=float64, chunksize=(31,), chunktype=numpy.ndarray>
scipy/_lib/_array_api.py:161: in array_namespace
    return array_api_compat.array_namespace(*api_arrays)
        api_arrays = [dask.array<array, shape=(31,), dtype=float64, chunksize=(31,), chunktype=numpy.ndarray>, array([1.])]
        arrays     = (dask.array<array, shape=(31,), dtype=float64, chunksize=(31,), chunktype=numpy.ndarray>, array([1.]))
scipy/_lib/array_api_compat/common/_helpers.py:587: in array_namespace
    raise TypeError(f"Multiple namespaces for array inputs: {namespaces}")
E   TypeError: Multiple namespaces for array inputs: {<module 'scipy._lib.array_api_compat.numpy' from '/home/runner/work/scipy/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>, <module 'scipy._lib.array_api_compat.dask.array' from '/home/runner/work/scipy/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/dask/array/__init__.py'>}
```

this suggests a bug in the test, it should be passing arrays with matching namespaces

```
___ TestResample.test_resample_methods[dask.array-polyphase-True-symmetric] ____
scipy/signal/tests/test_signaltools.py:1562: in test_resample_methods
    y_to = np.interp(t_to, t, x)
        corr       = dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>
        down       = 100
        ext        = True
        f_c        = 0.004975124378109453
        freq       = 40.0
        freqs      = dask.array<getitem, shape=(3, 1), dtype=float64, chunksize=(3, 1), chunktype=numpy.ndarray>
        g          = 1
        half_len   = 2010
        i          = 2
        max_rate   = 201
        method     = 'polyphase'
        padtype    = 'symmetric'
        polyargs   = {'padtype': 'symmetric', 'window': array([-7.12447662e-21, -9.15356661e-08, -1.84276451e-07, ...,
       -1.84276451e-07, -9.15356661e-08, -7.12447662e-21], shape=(4021,))}
        rate       = 100
        rate_to    = 49
        rates_to   = [49, 50, 51, 99, 100, 101, ...]
        rng        = RandomState(MT19937) at 0x7FC6CB95D040
        self       = <scipy.signal.tests.test_signaltools.TestResample object at 0x7fc75965e710>
        t          = dask.array<truediv, shape=(100,), dtype=float64, chunksize=(100,), chunktype=numpy.ndarray>
        t_to       = dask.array<truediv, shape=(49,), dtype=float64, chunksize=(49,), chunktype=numpy.ndarray>
        up         = 201
        window     = array([-7.12447662e-21, -9.15356661e-08, -1.84276451e-07, ...,
       -1.84276451e-07, -9.15356661e-08, -7.12447662e-21], shape=(4021,))
        x          = dask.array<array, shape=(100,), dtype=float64, chunksize=(100,), chunktype=numpy.ndarray>
        xp         = <module 'scipy._lib.array_api_compat.dask.array' from '/home/runner/work/scipy/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/dask/array/__init__.py'>
        y_resamp   = dask.array<getitem, shape=(201,), dtype=float64, chunksize=(201,), chunktype=numpy.ndarray>
        y_resamps  = dask.array<getitem, shape=(3, 201), dtype=float64, chunksize=(3, 201), chunktype=numpy.ndarray>
        y_to       = dask.array<getitem, shape=(201,), dtype=float64, chunksize=(201,), chunktype=numpy.ndarray>
        y_tos      = dask.array<mul, shape=(3, 201), dtype=float64, chunksize=(3, 201), chunktype=numpy.ndarray>
/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/dask/array/core.py:1766: in __array_function__
    return handle_nonmatching_names(func, args, kwargs)
        args       = (dask.array<truediv, shape=(49,), dtype=float64, chunksize=(49,), chunktype=numpy.ndarray>, dask.array<truediv, shape=...), chunktype=numpy.ndarray>, dask.array<array, shape=(100,), dtype=float64, chunksize=(100,), chunktype=numpy.ndarray>)
        func       = <function interp at 0x7fc7d69b28f0>
        handle_nonmatching_names = <function Array.__array_function__.<locals>.handle_nonmatching_names at 0x7fc6cbf472e0>
        kwargs     = {}
        module     = <module 'dask.array' from '/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/dask/array/__init__.py'>
        self       = dask.array<truediv, shape=(49,), dtype=float64, chunksize=(49,), chunktype=numpy.ndarray>
        types      = (<class 'dask.array.core.Array'>,)
/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/dask/array/core.py:1729: in handle_nonmatching_names
    warnings.warn(
E   FutureWarning: The `numpy.interp` function is not implemented by Dask array. You may want to use the da.map_blocks function or something similar to silence this warning. Your code may stop working in a future release.
```

this looks like we should be calling `np.interp` on a NumPy array first, then convert to `xp`

ev-br commented:
CI failures are unrelated.

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -4745,6 +4763,8 @@ def filtfilt(b, a, x, axis=-1, padtype='odd', padlen=None, method='pad',
     if edge > 0:
         # Slice the actual signal from the extended signal.
         y = axis_slice(y, start=edge, stop=-edge, axis=axis)
+        if is_torch(xp):
+            y = y.copy()    # XXX: torch chokes on negative strides
```
could we add a link to this into the source: https://github.com/pytorch/pytorch/issues/59786

lucascolley commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -372,25 +374,31 @@ def firwin(numtaps, cutoff, *, width=None, window='hamming', pass_zero=True,
     array([ 0.04890915,  0.91284326,  0.04890915])
 
     """
+    if isinstance(cutoff, int | float):
+        xp = np_compat
+    else:
+        xp = array_namespace(cutoff)
```
we can just use `xp = array_namespace(cutoff)` here and elsewhere, no?

ev-br replied:
Erm. I was under the impression that at least one array is obligatory, but apparently the scipy's version does not require that anymore. I was expecting that long-term we'd just use `array_namespace` from array_api_compat, but turns out we went the other way and are now adding features to it:

```
In [6]: from array_api_compat import array_namespace

In [7]: array_namespace(1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[7], line 1
----> 1 array_namespace(1)

File ~/.conda/envs/scipy-dev/lib/python3.12/site-packages/array_api_compat/common/_helpers.py:566, in array_namespace(api_version, use_compat, *xs)
    563         raise TypeError(f"{type(x).__name__} is not a supported array type")
    565 if not namespaces:
--> 566     raise TypeError("Unrecognized array input")
    568 if len(namespaces) != 1:
    569     raise TypeError(f"Multiple namespaces for array inputs: {namespaces}")

TypeError: Unrecognized array input

In [8]: from scipy._lib._array_api import array_namespace

In [9]: array_namespace(1)
Out[9]: <module 'scipy._lib.array_api_compat.numpy' from '/home/ev-br/repos/scipy/build-install/lib/python3.12/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>

```

lucascolley replied:
> but turns out we went the other way and are now adding features to it

Yep, this is by design since the very start! We need to wrap `array_namespace` to default to NumPy for array-likes.

lucascolley commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -426,26 +434,27 @@ def firwin(numtaps, cutoff, *, width=None, window='hamming', pass_zero=True,
     for left, right in bands:
-        h += right * sinc(right * m)
-        h -= left * sinc(left * m)
+        h += right * xpx.sinc(right * m, xp=xp)
+        h -= left * xp.sinc(left * m)
```
typo? Why aren't array-api-strict tests choking on this `xp.sinc`?

ev-br replied:
Good catch, thanks. Corrected. Seems to hint that test coverage can be improved in a follow-up.

lucascolley commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -623,30 +635,34 @@ def firwin2(numtaps, freq, gain, *, nfreqs=None, window='hamming',
     # Linearly interpolate the desired response on a uniform mesh `x`.
     x = np.linspace(0.0, nyq, nfreqs)
-    fx = np.interp(x, freq, gain)
+    fx = np.interp(x, np.asarray(freq), np.asarray(gain))   # XXX: xpx.interp
```
open an issue at xpx?

ev-br replied:
https://github.com/data-apis/array-api-extra/issues/193 and added a comment here.

lucascolley commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -1247,45 +1278,47 @@ def minimum_phase(h: np.ndarray,
-        w = np.arange(n_fft) * (2 * np.pi / n_fft * n_half)
-        H = np.real(fft(h, n_fft) * np.exp(1j * w))
+        w = xp.arange(n_fft, dtype=xp.float64) * (2 * xp.pi / n_fft * n_half)
+        I = xp.asarray(1j, dtype=xp.complex64 if w.dtype==xp.float32 else xp.complex128)
```
can eventually be simplified with https://github.com/data-apis/array-api/pull/848

ev-br replied:
We can actually simplify it today: `xp.exp(1j * w)` works with array api 2024.12 spec.

ev-br commented:
Review comments addressed.

lucascolley commented:
review time :) I haven't looked through the tests. Someone else is welcome to take that on if they want to. https://github.com/scipy/scipy/pull/22343#discussion_r2020202133 makes me hesitant to just yolo it

ev-br commented:
Read through the PR once again, removed several small pieces of cruft, and clean up dask skips: the majority needed an `np.asarray` around unconverted `freqs/freqz`, and one is still needed because `dask`'s fancy indexing is a bit limited.

DietBru commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -426,26 +432,27 @@ def firwin(numtaps, cutoff, *, width=None, window='hamming', pass_zero=True,
                              f'{pass_zero}')
     pass_zero = bool(operator.index(pass_zero))  # ensure bool-like
```
Side note: According to the docstr, `pass_zero in (True, False, 'bandpass', 'lowpass', 'highpass', 'bandstop')` needs to hold.
Couldn't the line 
```python
pass_zero = bool(operator.index(pass_zero))
``` 
be replaced with
```python
if isinstance(pass_zero, str):
   ...
elif pass_zero not in (True, False)):
    raise ValueError(...)
```
to make the code a little more accessible?

ev-br replied:
Great idea! Not sure what `...` would be though. Would you add a patch? (push to my branch or drop a patch here for me to apply or fix in a follow-up, whichever is easier).

ev-br replied:
Thank you! PR merged into this branch in https://github.com/scipy/scipy/pull/22343/commits/6c30b7d0ef8200c39c497b01a40eafe21d358139

DietBru commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -426,26 +432,27 @@ def firwin(numtaps, cutoff, *, width=None, window='hamming', pass_zero=True,
     pass_zero = bool(operator.index(pass_zero))  # ensure bool-like
 
-    pass_nyquist = bool(cutoff.size & 1) ^ pass_zero
+    pass_nyquist = bool(xp_size(cutoff) & 1) ^ pass_zero
```
Another matter of taste comment: I think
```suggestion
    pass_nyquist = (xp_size(cutoff) % 2 == 0) != pass_zero
```
is easier to comprehend (also due to ` numtaps % 2 == 0` being used in the line below).

ev-br replied:
Would certainly be more readable! However, should it be `== pass_zero` though?

```
In [11]: for x in (42, 43):
    ...:     for pass_zero in (True, False):
    ...:         old = bool(x & 1) ^ pass_zero
    ...:         new = (x % 2 == 0) != pass_zero
    ...:         print(x, pass_zero, old, new)
    ...: 
42 True True False
42 False False True
43 True False True
43 False True False
```

DietBru replied:
Of course :clown_face:! Fixed [here](https://github.com/ev-br/scipy/pull/22).

ev-br replied:
Thanks! Included in https://github.com/scipy/scipy/pull/22343/commits/6c30b7d0ef8200c39c497b01a40eafe21d358139

DietBru commented on scipy/signal/_signaltools.py:
```diff
@@ -2893,12 +2898,13 @@ def envelope(z: np.ndarray, bp_in: tuple[int | None, int | None] = (1, None), *,
         Z = sp_fft.fft(z)
     else:  # avoid calculating negative frequency bins for real signals:
-        Z = np.zeros_like(z, dtype=sp_fft.rfft(z.flat[:1]).dtype)
+        dt = sp_fft.rfft(xp.reshape(z, (-1,))[:1]).dtype
```
Another style comment: What do you think about
```suggestion
        dt = sp_fft.rfft(xp[..., :1]).dtype
```
It should work due `z = xp.moveaxis(z, axis, -1)`.

DietBru reviewed (commented):
I looked through this PR and it looks good to me, at least from the DSP side of things. I do not feel confident enough yet to make an assessment on the array API standard usage.

Most of my comments address programming style issues, which are of course a matter of personal taste. So feel free to only incorporate what you feel to fit.

ev-br commented on scipy/signal/tests/test_fir_filter_design.py:
```diff
@@ -330,110 +370,142 @@ def test_invalid_args(self):
+
+        freqs, response = freqz(np.asarray(taps), worN=2048)    # XXX convert freqz
+        assert_array_almost_equal(abs(xp.asarray(response)),
+                                  xp.asarray(freqs / np.pi), decimal=4)
```
No reason other than `freqs` being a numpy array here still. Not that it matters all that much though, they all are `math.pi` IIUC.

ev-br commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -1523,73 +1558,88 @@ def test_resample_methods(self, method, ext, padtype, xp):
-                    y_g = correlate1d(x, weights[::-1], mode='constant')
+                    x, weights = map(xp.asarray, (x, weights))
+                    flip = array_namespace(x).flip
+                    y_g = correlate1d(x, flip(weights), mode='constant')
```
This is an artifact of the testing system: the `xp` test fixture resolves to either a wrapped namespace or a bare namespace, depending on whether we test with or without `SCIPY_ARRAY_API` env variable set. If it is a bare namespace, it may be not entirely array api compatible. So this stanza makes sure we get the compatible version.

So,
- this is a temporary thing, only needed while the Array API support is behind a flag
- it should be only ever needed in tests, not code
- typically, I'd just assume array api compatibility and use this pattern only when testing finds an issue otherwise

ev-br commented on scipy/signal/tests/test_fir_filter_design.py:
```diff
@@ -505,112 +587,131 @@ def test_bad_args(self):
-        idx = np.logical_and(f > 0, f < a)
-        assert_array_almost_equal(Hmag[idx], np.ones_like(Hmag[idx]), decimal=3)
+        idx = (f > 0) & (f < a)
+        assert_array_almost_equal(Hmag[idx], xp.ones_like(Hmag[idx]), decimal=3)
```
No reason, just muscle memory. Done.

ev-br commented:
Thanks for the review @DietBru ! Addressed your comments in the last commit.

DietBru reviewed (approved):
LGTM, from the signal processing side. As said, I am not versed enough in the Array API to provide a useful opinion for all changes.

tylerjereddy commented:
@ev-br can you resolve the merge conflicts?

ev-br committed:
MAINT: signal: array API compat in several functions:

- envelope
- resample
- resample_poly

This completes the conversion of the 'filtering' rubic of scipy.signal (https://docs.scipy.org/doc/scipy/reference/signal.html#filtering)
While at it, convert

- remez
- firwin
- firwin2
- firls
- minimum_phase

This completes tests/test_fir_filter_design.py

Co-authored-by: Dietrich Brunn <Dietrich.Brunn@web.de>

ev-br committed:
MAINT: signal/ndimage: clean up delegators/support_alternative_backends

ev-br committed:
BUG: signal: fix firwin for non-numpy inputs

ev-br committed:
TST: signal: fix/expand firwin tests

ev-br committed:
MAINT: signal: remove a duplicate test

ev-br committed:
TST: signal: skip a test under jax

ev-br commented:
Rebased on main to resolve conflicts, and added a smoke test to hopefully address the concern  of https://github.com/scipy/scipy/pull/22343#discussion_r2020202133

ev-br committed:
BUG: signal: fix issues for the float32 default

rgommers commented:
Rebased to see if CI is still green after the recent submodule updates and test infra changes.

rgommers commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -372,25 +373,29 @@ def firwin(numtaps, cutoff, *, width=None, window='hamming', pass_zero=True,
     nyq = 0.5 * fs
 
-    cutoff = np.atleast_1d(cutoff) / float(nyq)
+    cutoff = xp.asarray(cutoff, dtype=xp_default_dtype(xp))
```
I think this should use `_asarray` (from `scipy._lib._array_api`) rather than `xp.asarray`?

lucascolley replied:
I don't think that is needed. `_asarray` may have been doing extra things in the past, but IIUC it is now only useful where we want to use `check_finite`, or maintain `order` or `subok` for NumPy:

https://github.com/scipy/scipy/blob/0762fee6adc973ddf07c4e7fe214f9733823e783/scipy/_lib/_array_api.py#L194-L214

lucascolley replied:
(or do both `asarray` and `array_namespace` in a one-liner, where there is no further use of `xp`)

ev-br replied:
Indeed, I frankly see no reason to invoke `_asarray` unless we need `check_finite`, which we're not using here.

ev-br replied:
> (or do both asarray and array_namespace in a one-liner, where there is no further use of xp)

That would need to call `array_namespace` twice because `xp_default_dtype(xp)`.

lucascolley replied:
> > (or do both asarray and array_namespace in a one-liner, where there is no further use of xp)
>
> That would need to call array_namespace twice because xp_default_dtype(xp).

Indeed, I wasn't talking about this specific case, but rather why `_asarray` _might_ be useful elsewhere

rgommers commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -573,11 +577,14 @@ def firwin2(numtaps, freq, gain, *, nfreqs=None, window='hamming',
 
     """
+    xp = array_namespace(freq, gain)
+    freq, gain = xp.asarray(freq), xp.asarray(gain)
```
same here, `_asarray`

rgommers commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -822,6 +831,12 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
 
     """
+    xp = array_namespace(bands, desired, weight)
+    bands = np.asarray(bands)
```
Good to keep this as `np.` for now. Can be converted to the `lazy_apply` mechanism later, once the dust settles and we've drawn a conclusion on whether we like that and will keep it unchanged.

ev-br replied:
`remez` invokes a C extension, so its inputs need to get converted to numpy on way or another. Whether long-term we want it via `np.asarray` or some other method is TBD I guess, but today this is the way.

lucascolley replied:
yep, that would be https://data-apis.org/array-api-extra/generated/array_api_extra.lazy_apply.html with `as_numpy=True` down the line

rgommers commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -1232,13 +1252,16 @@ def minimum_phase(h: np.ndarray,
-    if np.iscomplexobj(h):
+    xp = array_namespace(h)
+
+    h = xp.asarray(h)
```
`_asarray`

rgommers commented on scipy/signal/_signaltools.py:
```diff
@@ -3926,7 +3929,9 @@ def resample_poly(x, up, down, axis=0, window=('kaiser', 5.0),
-    x = np.asarray(x)
+    xp = array_namespace(x)
+
+    x = xp.asarray(x)
```
`_asarray`

rgommers commented on scipy/signal/_signaltools.py:
```diff
@@ -4841,6 +4848,8 @@ def filtfilt(b, a, x, axis=-1, padtype='odd', padlen=None, method='pad',
         # Slice the actual signal from the extended signal.
         y = axis_slice(y, start=edge, stop=-edge, axis=axis)
+        if is_torch(xp):
+            y = y.copy()    #  pytorch/pytorch#59786 : no negative strides in pytorch
```
It's unclear to me why this is needed; the slicing is done in `axis_slice` above, once we get here the function is done already.

ev-br replied:
There's an `axis_revese` just above which leads to

```
4848 	    # Reverse y.
4849 	    y = axis_reverse(y, axis=axis)
4850 	
4851 ->	    if edge > 0:
4852 	        # Slice the actual signal from the extended signal.
4853 	        y = axis_slice(y, start=edge, stop=-edge, axis=axis)
4854 	    #    if is_torch(xp):
4855 	    #        y = y.copy()    #  pytorch/pytorch#59786 : no negative strides in pytorch
4856 	
(Pdb) p y.strides
(-8,)
```
and then `xp.asarray(y)` fails if `xp` is `torch`.

rgommers replied:
`xp.flip` should be functional, it won't modify in-place for any library (that would be a serious bug). So maybe a useful cleanup for a future PR?

ev-br replied:
Ah, then it does not get any simpler than what's here now. `axis_reverse` does things without copying, basically a `slice(None, None, -1)`, and it's only torch which requires an extra copy. One could stuff a `.copy` into `axis_reverse`  but then it returns a copy or a view depending on a backend. Which is a no-no if we can help it, even for private helpers.

rgommers replied:
I didn't say that `flip` has to copy! That would be bad indeed. It's allowed to return a view, that is still functional.

```python
>>> x = np.arange(100_000)
>>> %timeit x.copy()
33.2 μs ± 1.48 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
>>> %timeit np.flip(x)
426 ns ± 3.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

`flip` should be doing exactly what you want here.

rgommers replied:
Minor issue of semantics: you asked about "is in-place", which to me means modifying existing array data (an in-place operation), rather than "can it return a view".

rgommers commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -1419,64 +1424,89 @@ def test_basic(self, xp):
             signal.resample(tsig, num, domain='time'),
             atol=1e-9)
 
+    @skip_xp_backends("jax.numpy", reason="XXX: immutable arrays")
```
nit: `XXX` can be deleted, looks like debug comment (only suggest fixing in case you're pushing another commit) 

Applies in a few more places

ev-br replied:
I plan to have a cleanup pass across the whole `test_signaltools.py`, once other scipy.signal PRs land (this part of scipy is _very_ intertwined, esp in testing).

rgommers commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -1519,9 +1549,13 @@ def test_resample_methods(self, method, ext, padtype, xp):
+                freq = float(freqs[i, 0])
                 if freq >= 0.5 * rate_to:
-                    y_to.fill(0.)  # mostly low-passed away
+                    #y_to.fill(0.)  # mostly low-passed away
```
line can be deleted

rgommers reviewed (commented):
LGTM, thanks @ev-br. A few tiny comments and one substantial one: unless I missed something in how we do things, I believe conversion and input validation of array-likes has to be done with `_asarray` from `scipy._lib._array_api` rather than with `xp.asarray`.

ev-br commented:
Thanks for taking a look @rgommers. 
I don't see what `_array` buys us compared to `xp.asarray` here, mind explaining? 
And if there is a reason, I guess it's best changed across the whole `scipy.signal`, as currently both `ndimage` and `signal` use just `xp.asarray` throughout.
Other small comments I'd prefer to lump into a single sweep across `scipy.signal` tests. Which would be way more effective to do after https://github.com/scipy/scipy/pull/22911 and https://github.com/scipy/scipy/pull/22914 land, IOW when we'll have all of `test_signaltools.py`, `test_filter_design.py` and `test_fir_filter_design.py` done.
CI failures look unrelated.

rgommers commented:
> I don't see what `_array` buys us compared to `xp.asarray` here, mind explaining?

Nothing anymore, looks like my knowledge on the right code pattern for this was outdated. As Lucas said above: 

*`_asarray` may have been doing extra things in the past, but IIUC it is now only useful where we want to use `check_finite`, or maintain `order` or `subok` for NumPy*

> Other small comments I'd prefer to lump into a single sweep across `scipy.signal` tests.

Sure, SGTM.

rgommers reviewed (approved):
CI failures are unrelated and all seems well, so in it goes. Thanks @ev-br & reviewers!
